### PR TITLE
feature/cookie crumb and retries

### PR DESCRIPTION
**Pull request by Jshen123**
Credits to @ValueRaider  and @psychoz971  for doing bulk of the work
- https://github.com/ranaroussi/yfinance/pull/1657/files
- https://github.com/ranaroussi/yfinance/issues/1729#issuecomment-1790830481

I have combined both their solutions in this PR.  The PR should allow you to get `info` leveraging cookie and crumb

The solution is not 100% reliable (from time to time, your API call will encounter 401 error) so I have also added a retry mechanism. From my testing, I never needed more than 5 retries but I have set max_retries to 10 just to be on the safe side.

### CODE REVIEW

File: yfinance/data.py [hunk 1]
```diff
@@ -71,6 +79,49 @@ def _get_proxy(self, proxy):
         return proxy
 
     def get_raw_json(self, url, user_agent_headers=None, params=None, proxy=None, timeout=30):
-        response = self.get(url, user_agent_headers=user_agent_headers, params=params, proxy=proxy, timeout=timeout)
-        response.raise_for_status()
+        retry = 0
+        response = None
+
+        while response is None or (retry < self.max_retries and 400 <= response.status_code):
```
Comment (Jshen123) [inline yfinance/data.py:85]: retry in the event the API call fails

**Comment by Jshen123** [comment]
FYI, a lot of automatic formatting changes from prettier

File: yfinance/scrapers/quote.py [hunk 2]
```diff
@@ -595,28 +603,37 @@ def _fetch(self, proxy):
         if self._already_fetched:
             return
         self._already_fetched = True
-        modules = ['financialData', 'quoteType', 'defaultKeyStatistics', 'assetProfile', 'summaryDetail']
+        modules = ("summaryDetail,"
```
Comment (Jshen123) [inline yfinance/scrapers/quote.py:606]: changed the shape of `modules` to work with v10

File: yfinance/data.py [hunk 3]
```diff
@@ -71,6 +79,49 @@ def _get_proxy(self, proxy):
...
+            response = self.get(url, user_agent_headers=user_agent_headers, params=params,
+                                proxy=proxy, timeout=timeout)
+            print(retry)
+            retry += 1
+
+        if 400 <= response.status_code:
+            utils.warn_for_status(response)
```
Comment (Jshen123) [inline yfinance/data.py:92]: Replaced `raise` with `warn` so it doesn't throw the entire bulk operation in the event that some of the calls failed

**Comment by Jshen123** [comment]
Hi @ValueRaider, let me know your thoughts on this :)

**Comment by jkoestner** [comment]
I didn't extensively test...your method did work for the .info method, however it seemed like it broke other methods such as .download

**Comment by ValueRaider** [comment]
@jkoestner Where on earth are you? I mean that literally, as this branch works for me in Europe and I didn't experience your problems.

**Comment by jkoestner** [comment]
> @jkoestner Where on earth are you? I mean that literally, as this branch works for me in Europe and I didn't experience your problems.

I'm in the US.

I didn't do a deep dive but thought that using v10 may have restructured the output and could be reason for below response.

image

**Comment by Jshen123** [comment]
Hi @jkoestner, do you happen to be on VPN?

I was able to repro your issue while I'm on VPN (regardless of region). Otherwise, it works as expected.

**Comment by ValueRaider** [comment]
What does this take from #1657? I'm only seeing ...

**Comment by Jshen123** [comment]
> What does this take from #1657? I'm only seeing `psychoz971` solution which has a 50% success rate [#1729 (comment)](https://github.com/ranaroussi/yfinance/issues/1729#issuecomment-1793548822).

Oh, I was under the assumption that `get_yahoo_cookie` and `get_yahoo_crumb` was inspired by your PR. If not, this is entire based on @psychoz871

Note on the success rate. I think with retry, the success rate increases significantly. For me, it hasn't failed yet

**Comment by ValueRaider** [comment]
PR has 2 strategies now.

**Comment by ValueRaider** [comment]
Copy over my patch for missing keys https://github.com/ranaroussi/yfinance/pull/1657/commits/2f241e60306436ab76c2d14f0627c19f59795aeb

**Comment by Jshen123** [comment]
Refer to https://github.com/ranaroussi/yfinance/pull/1736 for updated PR

**Comment by kfwalther** [comment]
I have this PR checked out and it's working for me. Why did it get closed?

**Comment by Jshen123** [comment]
> I have this PR checked out and it's working for me. Why did it get closed?

Hi @kfwalther, changes have now been incorporated in https://github.com/ranaroussi/yfinance/pull/1657

**Pull request closed without merging**